server/lib/actors/hot.go

```diff
@@ -0,0 +1,741 @@
+			//log.Printf("hots: %v %v %v %v %v", e, from, to, yue.DateByUUID(yuep.UUID(e.Id)), e.within(from, to))
+			return e.within(from, to)
+		}).sort(proto.TopicListRequest_Hot).truncate(DEFAULT_HOTBUCKET_SIZE).markUpdate(now)
+		next.flames[i] = next.risings[i].filter(func (e *HotEntry) bool {
+			return e.within(from, to) && e.flamy()
+		}).sort(proto.TopicListRequest_Flame).truncate(DEFAULT_HOTBUCKET_SIZE).markUpdate(now)
+	}
+
+	//log.Printf("mergeFetchResult2")
+	//week/alltime hot and flame. merge query results and fetchCaches
+	for ; i < (len(next.risings) + len(a.queryDurations)); i++ {
+		index := i - len(next.risings)
+		next.hots[i] = a.hotQueryCache[index].Load().(*hotbucket)
+		next.flames[i] = a.flameQueryCache[index].Load().(*hotbucket)
+		from := to.Add(-1 * a.queryDurations[index])
+		var filter func (e *HotEntry) bool = nil 
+		if index == 0 {
+			filter = func (e *HotEntry) bool {
+				return e.within(from, to)
+			}
+		}
+		for _, c := range next.caches {
+			//if filter == nil {
+			//	log.Printf("alltime hot: %v %v %v", c.fetchAt, next.hots[i].lastUpdate, c.results)
+			//}
+			if c.fetchAt.UnixNano() > next.hots[i].lastUpdate.UnixNano() {
+				next.hots[i].addResultsWithFilter(c.results, filter)
+			}
+			if c.fetchAt.UnixNano() > next.flames[i].lastUpdate.UnixNano() {
+				next.flames[i].addResultsWithFilter(c.results, filter)
+			}
+		}
+		//hots just reorder. set position of markUpdate prior to truncate is important because truncate create new instance,
+		//which means markupdate after truncate does not propagate to original data (a.***QueryCache)
+		next.hots[i] = next.hots[i].markUpdate(now).sort(proto.TopicListRequest_Hot).truncate(DEFAULT_HOTBUCKET_SIZE)
+		//flames will be filtered, because each of its element may be non flamy. (also be careful about position of markUpdate)
+		next.flames[i] = next.flames[i].markUpdate(now).filter(func (e *HotEntry) bool {
+			return e.flamy()
+		}).sort(proto.TopicListRequest_Flame).truncate(DEFAULT_HOTBUCKET_SIZE).markUpdate(now)
+	}
+}
+
+func (a *HotActor) update(t time.Time) (err error) {
+	defer func() {
+		if err == nil {
+			a.lastQuery = t
+		}
+	}()
+	var results []FetchResult
+	if results, err = a.fetcher(proto.TopicListRequest_Rising, a.lastQuery, t, a.locale); err != nil {
+		return err
+	}
+	prev := a.store.Load().(*hotbucketStore)
+	next := prev.update(t, results)
+	a.mergeFetchResult(next, t)
+	a.store.Store(next)
+	go a.save(a.lastQuery, prev)
+	return nil
+}
+
+func (a *HotActor) save(at time.Time, bs *hotbucketStore) error {
+	return a.persist(at, bs)
+}
+
+func (a *HotActor) destroy() {
+	//please add shutdown stuff
+}
+
+func (a *HotActor) restore() error {
+	//TODO: restore summeries and unitCaches from database or filesystem.
+	return nil
+}
+
+func (a *HotActor) Available() bool {
+	return a.available
+}
+
+func (a *HotActor) Query(btyp proto.TopicListRequest_BucketType, typ proto.TopicListRequest_QueryType, 
```

Comment:
__[gometalinter]__
* [gotype] undeclared name: proto
* [golint] exported method HotActor.Query should have comment or be unexported